camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java

```diff
@@ -0,0 +1,173 @@
+            String uri,
+            KameletComponent component,
+            String templateId,
+            String routeId,
+            Map<String, Object> kameletProperties) {
+
+        super(uri, component);
+
+        ObjectHelper.notNull(templateId, "template id");
+        ObjectHelper.notNull(routeId, "route id");
+        ObjectHelper.notNull(kameletProperties, "kamelet properties");
+
+        this.templateId = templateId;
+        this.routeId = routeId;
+        this.kameletProperties = Collections.unmodifiableMap(kameletProperties);
+        this.kameletUri = "direct:" + routeId;
+    }
+
+    @Override
+    public KameletComponent getComponent() {
+        return (KameletComponent) super.getComponent();
+    }
+
+    public String getTemplateId() {
+        return templateId;
+    }
+
+    public String getRouteId() {
+        return routeId;
+    }
+
+    public Map<String, Object> getKameletProperties() {
+        return kameletProperties;
+    }
+
+    @Override
+    public Producer createProducer() throws Exception {
+        return new KameletProducer();
+    }
+
+    @Override
+    public Consumer createConsumer(Processor processor) throws Exception {
+        Consumer answer = new KemeletConsumer(processor);
+        configureConsumer(answer);
+        return answer;
+    }
+
+    @Override
+    protected void doInit() throws Exception {
+        super.doInit();
+
+        getComponent().onEndpointAdd(this);
+    }
+
+    // *********************************
+    //
+    // Helpers
+    //
+    // *********************************
+
+    private class KemeletConsumer extends DefaultConsumer {
+        private volatile Endpoint endpoint;
+        private volatile Consumer consumer;
+
+        public KemeletConsumer(Processor processor) {
+            super(KameletEndpoint.this, processor);
+        }
+
+        @Override
+        protected void doStart() throws Exception {
+            endpoint = getCamelContext().getEndpoint(kameletUri);
+            consumer = endpoint.createConsumer(getProcessor());
+
+            ServiceHelper.startService(endpoint, consumer);
+            super.doStart();
+        }
+
+        @Override
+        protected void doStop() throws Exception {
+            ServiceHelper.stopService(consumer, endpoint);
+            super.doStop();
+        }
+    }
+
+    private class KameletProducer extends DefaultAsyncProducer {
+        private volatile Endpoint endpoint;
+        private volatile AsyncProducer producer;
+
+        public KameletProducer() {
+            super(KameletEndpoint.this);
+        }
+
+        @Override
+        public boolean process(Exchange exchange, AsyncCallback callback) {
+            if (producer != null) {
+                return producer.process(exchange, callback);
+            } else {
+                callback.done(true);
+                return true;
+            }
+        }
+
+        @Override
+        protected void doStart() throws Exception {
```

Comment:
You can maybe get the endpoint and create producer in doInit and the same for the consumer

Reply:
I can probably do it for the producer but the consumer needs the processor so I can't eagerly create it right ?